File: smtk/session/vtk/testing/python/smtkReadFile.py
# =============================================================================
#
#  Copyright (c) Kitware, Inc.
#  All rights reserved.
#  See LICENSE.txt for details.
#
#  This software is distributed WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#  PURPOSE.  See the above copyright notice for more information.
#
# =============================================================================
import os
import smtk
import smtk.model
import smtk.session.vtk
import smtk.testing
import sys


class TestVTKSession(smtk.testing.TestCase):

    def testReadSMTK(self):
        smtk_file = os.path.join(
            smtk.testing.DATA_DIR, 'model', '3d', 'exodus', 'disk_out_ref.smtk')
        print('smtk_file', smtk_file)

        # Import the input file
        readOp = smtk.session.vtk.LegacyRead.create()
        readOp.parameters().find('filename').setValue(smtk_file)
        readRes = readOp.operate()
        self.assertEqual(readRes.find('outcome').value(0),
                         int(smtk.operation.Operation.SUCCEEDED),
                         'read model failed')

        # Access the resource
        resource = smtk.model.Resource.CastTo(
            readRes.find('resourcesCreated').value())
        models = resource.findEntitiesOfType(int(smtk.model.MODEL_ENTITY))
        self.assertEqual(len(models), 1)
        model = models[0]

        self.assertTrue(model.isValid(), 'Imported model not valid')

        # Verify that the model is marked as discrete
        self.assertEqual(
            model.geometryStyle(), smtk.model.DISCRETE,
            'Expected discrete model, got {gs}'.format(gs=model.geometryStyle()))


if __name__ == '__main__':
    smtk.testing.process_arguments()
    smtk.testing.main()

File: smtk/session/vtk/testing/cxx/smtkReadFile.cxx
//=========================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//=========================================================================

#include "smtk/AutoInit.h"
#include "smtk/PublicPointerDefs.h"
#include "smtk/attribute/Attribute.h"
#include "smtk/attribute/FileItem.h"
#include "smtk/attribute/IntItem.h"
#include "smtk/attribute/ResourceItem.h"
#include "smtk/model/EntityRef.h"
#include "smtk/model/EntityTypeBits.h"
#include "smtk/model/Model.h"

#include "smtk/session/vtk/Resource.h"
#include "smtk/session/vtk/operators/LegacyRead.h"

#include "smtk/common/testing/cxx/helpers.h"

#include <iostream>
#include <string>

int main(int argc, char* argv[])
{
  // basic check info
  if (argc < 2)
  {
    std::cout << "Must provide input file as argument" << std::endl;
    return 1;
  }
  std::string modelPath(argv[1]);

  smtk::session::vtk::Resource::Ptr resource;
  smtk::model::Entity::Ptr modelEntity;

  {
    // Create a read operator
    smtk::session::vtk::LegacyRead::Ptr readOp = smtk::session::vtk::LegacyRead::create();
    if (!readOp)
    {
      std::cerr << "No read operator\n";
      return 1;
    }

    // Set the file path
    std::cout << "readOp parameters: " << readOp->parameters() << std::endl;
    readOp->parameters()->findFile("filename")->setValue(modelPath);

    // Execute the operation
    smtk::operation::Operation::Result readOpResult = readOp->operate();

    // Retrieve the resulting resource
    smtk::attribute::ResourceItemPtr resourceItem =
      std::dynamic_pointer_cast<smtk::attribute::ResourceItem>(
        readOpResult->findResource("resourcesCreated"));

    resource = std::dynamic_pointer_cast<smtk::session::vtk::Resource>(resourceItem->value());

    if (!resource)
    {
      std::cerr << "No resource\n";
      return 1;
    }

    // Retrieve the resulting model
    smtk::model::Models models =
      resource->entitiesMatchingFlagsAs<smtk::model::Models>(smtk::model::MODEL_ENTITY, false);

    std::cout << "found " << models.size() << " models" << std::endl;
    if (models.empty())
      return 1;

    modelEntity = models[0].entityRecord();

    // Test for success
    if (
      readOpResult->findInt("outcome")->value() !=
      static_cast<int>(smtk::operation::Operation::Outcome::SUCCEEDED))
    {
      std::cerr << "Read operator failed\n";
      return 1;
    }
  }

  auto model = modelEntity->referenceAs<smtk::model::Model>();

  // Check model validity
  test(model.isValid(), "Invalid model");

  // Check model geometry style
  test(model.geometryStyle() == smtk::model::DISCRETE, "Incorrect geometry style (not discrete)");

  return 0;
}
